src/Microsoft.ML/Runtime/Internal/Tools/GeneratorUtils.cs

```diff
@@ -0,0 +1,467 @@
+                fullTypeName = fullTypeName.Substring(0, bracketIndex);
+            }
+
+            // When the type is nested, the names of the outer types are concatenated with a '+'.
+            var nestedNames = fullTypeName.Split('+');
+            var baseName = nestedNames[0];
+
+            // We currently only handle generic types in the outer most class, support for generic inner classes
+            // can be added if needed.
+            int backTickIndex = baseName.LastIndexOf('`');
+            int dotIndex = baseName.LastIndexOf('.');
+            Contracts.Assert(dotIndex >= 0);
+            if (backTickIndex < 0)
+                name += baseName.Substring(dotIndex + 1);
+            else
+            {
+                name += baseName.Substring(dotIndex + 1, backTickIndex - dotIndex - 1);
+                Contracts.AssertValue(genericTypes);
+                if (genericTypes != null)
+                {
+                    foreach (var genType in genericTypes)
+                    {
+                        var splitNames = genType.FullName.Split('+');
+                        if (splitNames[0].LastIndexOf('.') >= 0)
+                            splitNames[0] = splitNames[0].Substring(splitNames[0].LastIndexOf('.') + 1);
+                        name += string.Join("", splitNames);
+                    }
+                }
+            }
+
+            for (int i = 1; i < nestedNames.Length; i++)
+                name += nestedNames[i];
+
+            Contracts.Assert(typesSymbolTable.Select(kvp => kvp.Value).All(str => string.Compare(str, name) != 0));
+            typesSymbolTable[type.FullName] = name;
+            return name;
+        }
+
+        public static void GenerateSummary(IndentingTextWriter writer, string summary)
+        {
+            if (string.IsNullOrEmpty(summary))
+                return;
+            writer.WriteLine("/// <summary>");
+            foreach (var line in summary.Split(new[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries))
+                writer.WriteLine($"/// {line}");
+            writer.WriteLine("/// </summary>");
+        }
+
+        public static void GenerateHeader(IndentingTextWriter writer)
+        {
+            writer.WriteLine("//------------------------------------------------------------------------------");
+            writer.WriteLine("// <auto-generated>");
+            writer.WriteLine("//     This code was generated by a tool.");
+            writer.WriteLine("//");
+            writer.WriteLine("//     Changes to this file may cause incorrect behavior and will be lost if");
+            writer.WriteLine("//     the code is regenerated.");
+            writer.WriteLine("// </auto-generated>");
+            writer.WriteLine("//------------------------------------------------------------------------------");
+            //writer.WriteLine($"// This file is auto generated. To regenerate it, run: {_regenerate}");
```

Comment:
>//writer.WriteLine($"// This file is auto generated. To regenerate it, run: {_regenerate}"); [](start = 12, length = 92)

We might prefer to avoid checking in commented out lines of code.